fs/shm/shmfs.c

```diff
@@ -0,0 +1,359 @@
+        {
+          delete_shm_object(inode->i_private);
+          inode->i_private = NULL;
+        }
+
+      inode_unlock();
+    }
+
+  return ret;
+}
+
+/****************************************************************************
+ * Name: shmfs_close
+ ****************************************************************************/
+
+static int shmfs_close(FAR struct file *filep)
+{
+  /* Release the shmfs object. The object gets deleted if no-one has
+   * reference to it (either mmap or open file) and the object has been
+   * unlinked
+   */
+
+  return shmfs_release(filep->f_inode);
+}
+
+/****************************************************************************
+ * Name: shmfs_truncate
+ ****************************************************************************/
+
+static int shmfs_truncate(FAR struct file *filep, off_t length)
+{
+  FAR struct shmfs_object_s *object;
+  int ret;
+
+  ret = inode_lock();
+  if (ret >= 0)
+    {
+      object = filep->f_inode->i_private;
+      if (!object)
+        {
+          filep->f_inode->i_private = alloc_shm_object(length);
+        }
+      else if (object->length != length)
+        {
+          /* This doesn't support resize */
+
+          ret = -EINVAL;
+        }
+
+      inode_unlock();
+    }
+
+  return ret;
+}
+
+/****************************************************************************
+ * Name: shmfs_unlink
+ ****************************************************************************/
+#ifndef CONFIG_DISABLE_PSEUDOFS_OPERATIONS
+static int shmfs_unlink(FAR struct inode *inode)
+{
+  if (inode->i_crefs <= 1)
+    {
+      delete_shm_object(inode->i_private);
+      inode->i_private = NULL;
+    }
+
+  return OK;
+}
+#endif
+
+/****************************************************************************
+ * Name: shmfs_mmap
+ ****************************************************************************/
+
+static int shmfs_mmap(FAR struct file *filep, FAR struct mm_map_entry_s *map)
+{
+  FAR struct shmfs_object_s *object;
+  int ret = -EINVAL;
+
+  DEBUGASSERT(filep->f_inode != NULL);
+
+  /* We don't support offset at the moment, just mapping the whole object */
+
+  if (map->offset != 0)
+    {
+      return ret;
+    }
+
+  /* Keep the inode when mmapped, increase refcount */
+
+  ret = inode_addref(filep->f_inode);
+  if (ret >= 0)
+    {
+      object = (FAR struct shmfs_object_s *)filep->f_inode->i_private;
+
+      /* object may be null if the shm has not been ftruncated */
+
+      if (object)
+        {
+#ifdef CONFIG_BUILD_KERNEL
+          FAR struct tcb_s        *tcb   = nxsched_self();
+          FAR struct task_group_s *group = tcb->group;
+          FAR uintptr_t *pages =
+            (FAR uintptr_t *)&object->paddr[0];
+          uintptr_t                mapaddr;
+          unsigned int             npages;
+
+          /* Find a free vaddr space that satisfies length */
+
+          mapaddr = (uintptr_t)shm_alloc(group, 0, object->length);
```

Comment:
This is a bug by me, the return value can be NULL, so it needs to be tested here.

Reply:
added a check